### PR TITLE
AST/IRGen: Accept `@_weakLinked` on import decls to force weak linkage of symbols from a module

### PR DESCRIPTION
**Pull request by tshortli**
The effect of declaring an import `@_weakLinked` is to treat every declaration from the module as if it were declared with `@_weakLinked`. This is useful in environments where entire modules may not be present at runtime. Although it is already possible to instruct the linker to weakly link an entire dylib, a Swift attribute provides a way to declare intent in source code and also opens the door to diagnostics and other compiler behaviors that depend on knowing that all the module's symbols will be weakly linked.

The `@_weakLinked` attribute on imports is not transitive; if module A imports module B which `@_weakLinked` imports module C, it does not follow that A should weakly link symbols from C. Just because B needs to operate on runtimes where C is absent does not mean A needs to also operate on runtimes where C is absent. Therefore the `@_weakLinked` attribute is not serialized in binary modules nor is it printed in swiftinterface files.

Resolves rdar://96098097.

### CODE REVIEW

**Comment by tshortli** [comment]
@swift-ci please smoke test

**Comment by tshortli** [comment]
@swift-ci please smoke test

**Comment by tshortli** [comment]
@swift-ci please smoke test

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please test

**Comment by tshortli** [comment]
@swift-ci please smoke test

File: lib/AST/Module.cpp [hunk 1]
```diff
@@ -2599,6 +2618,15 @@ bool ModuleDecl::isImportedAsSPI(Identifier spiGroup,
   return importedSPIGroups.count(spiGroup);
 }
 
+bool ModuleDecl::isImportedAsWeakLinked(const Decl *targetDecl) const {
+  const auto *declaringModule = targetDecl->getModuleContext();
+  for (auto file : getFiles()) {
+    if (file->importsModuleAsWeakLinked(declaringModule))
```
Comment (nkcsgexi) [inline lib/AST/Module.cpp:2624]: Should we diagnose some source files use `@ _weakLinked import Foo` but some others use `import Foo`?

File: lib/AST/Module.cpp [hunk 2]
```diff
@@ -2574,6 +2574,25 @@ bool SourceFile::isImportedAsSPI(const ValueDecl *targetDecl) const {
...
+  for (auto &import : *Imports) {
+    if (!import.options.contains(ImportFlags::WeakLinked))
+      continue;
+
+    const ModuleDecl *importedModule = import.module.importedModule;
+    if (module == importedModule)
+      return true;
+
+    // Also check whether the target module is actually the underlyingClang
+    // module for this @_weakLinked import.
```
Comment (nkcsgexi) [inline lib/AST/Module.cpp:2587]: Can we add a test case for the Swift overlay + underlying Clang module case?

**Review by nkcsgexi — COMMENTED** [review]
This is awesome! Just some nitpicks.

File: test/IRGen/weaklinked_import.swift [hunk 3]
```diff
@@ -0,0 +1,159 @@
+// RUN: %empty-directory(%t)
```
Comment (nkcsgexi) [inline test/IRGen/weaklinked_import.swift:1]: Could you also add a test for adding `@_weakLinked import Foo` where `Foo` is a transitive dependency? APIs from `Foo` are not directly used by the source file, but a direct dependency `Bar` of the source file has some APIs that trigger inlining `Foo`'s APIs. Those APIs should be weakly linked.

File: lib/AST/Module.cpp (hunk 1 above)
Comment (tshortli) [reply lib/AST/Module.cpp:2624]: Yes, I'm planning to do that in an upcoming PR.

File: lib/AST/Module.cpp (hunk 2 above)
Comment (tshortli) [reply lib/AST/Module.cpp:2587]: This is tested by the `weaklinked_import_clang` test case (I discovered needing this logic while writing it).

File: test/IRGen/weaklinked_import.swift (hunk 3 above)
Comment (tshortli) [reply test/IRGen/weaklinked_import.swift:1]: Good idea, I'll add that test.

File: lib/AST/Module.cpp (hunk 2 above)
Comment (nkcsgexi) [reply lib/AST/Module.cpp:2587]: I see. Thanks!

File: include/swift/SIL/SILFunction.h [hunk 4]
```diff
@@ -64,6 +64,11 @@ enum IsDistributed_t {
   IsNotDistributed,
   IsDistributed,
 };
+enum IsWeakImported_t {
```
[the remark this answers is not in the document]
Comment (tshortli) [reply include/swift/SIL/SILFunction.h:67]: I can make this one `uint8_t` but I was following the pattern of the enums above. I guess the `PerformanceConstraints` enum below already makes this file inconsistent.

File: test/IRGen/weaklinked_import.swift (hunk 3 above)
Comment (tshortli) [reply test/IRGen/weaklinked_import.swift:1]: This was a good catch. It turns out that the logic in this PR does not properly change the linkage of symbol references inlined from other modules. I think it's going to require some deeper investigation so I'd like to tackle it in another PR.

Comment (nkcsgexi) [reply test/IRGen/weaklinked_import.swift:1]: Sounds good!

File: include/swift/SIL/SILFunction.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (tshortli) [reply include/swift/SIL/SILFunction.h:67]: This actually makes it less convenient to store in a bitfield - I'm going to keep as-is.

**Comment by tshortli** [comment]
@swift-ci please test

File: test/IRGen/weaklinked_import.swift (hunk 3 above)
Comment (tshortli) [reply test/IRGen/weaklinked_import.swift:1]: I added a short test case that documents that this currently doesn't work.

**Review by nkcsgexi — APPROVED** [review]

**Pull request merged**